src/algorithm/model.hxx

```diff
@@ -468,67 +468,98 @@ namespace pinocchio
-    for(GeometryObjectVector::const_iterator it = input_geom_model.geometryObjects.begin();
-        it != input_geom_model.geometryObjects.end(); ++it)
-    {
-      const GeometryModel::GeometryObject & geom = *it;
-      
-      const JointIndex joint_id_in_input_model = geom.parentJoint;
-      _PINOCCHIO_CHECK_INPUT_ARGUMENT_2((joint_id_in_input_model < (JointIndex)input_model.njoints),
-                                        "Invalid joint parent index for the geometry with name " + geom.name);
-      const std::string & parent_joint_name = input_model.names[joint_id_in_input_model];
-      
-      JointIndex reduced_joint_id = (JointIndex)-1;
-      typedef typename Model::SE3 SE3;
-      SE3 relative_placement = SE3::Identity();
-      if(reduced_model.existJointName(parent_joint_name))
-      {
-        reduced_joint_id = reduced_model.getJointId(parent_joint_name);
-      }
-      else // The joint is now a frame
-      {
-        const FrameIndex reduced_frame_id = reduced_model.getFrameId(parent_joint_name);
-        reduced_joint_id = reduced_model.frames[reduced_frame_id].parent;
-        relative_placement = reduced_model.frames[reduced_frame_id].placement;
+  template <typename Scalar, int Options,
+            template <typename, int> class JointCollectionTpl,
+            typename ConfigVectorType>
+  void buildReducedModel(
+      const ModelTpl<Scalar, Options, JointCollectionTpl> &input_model,
+      const GeometryModel &input_geom_model,
+      const std::vector<JointIndex> &list_of_joints_to_lock,
+      const Eigen::MatrixBase<ConfigVectorType> &reference_configuration,
+      ModelTpl<Scalar, Options, JointCollectionTpl> &reduced_model,
+      GeometryModel &reduced_geom_model) {
+
+    const std::vector<GeometryModel> temp_input_geoms { input_geom_model };
+    std::vector<GeometryModel> temp_reduced_geom_models;
+
+    buildReducedModel(input_model, temp_input_geoms, list_of_joints_to_lock,
+                      reference_configuration, reduced_model,
+                      temp_reduced_geom_models);
+    reduced_geom_model = temp_reduced_geom_models.front();
+  }
+
+  template <typename Scalar, int Options,
+            template <typename, int> class JointCollectionTpl,
+            typename ConfigVectorType>
+  void buildReducedModel(
+      const ModelTpl<Scalar, Options, JointCollectionTpl> &input_model,
+      const std::vector<GeometryModel> &list_of_geom_models,
+      const std::vector<JointIndex> &list_of_joints_to_lock,
+      const Eigen::MatrixBase<ConfigVectorType> &reference_configuration,
+      ModelTpl<Scalar, Options, JointCollectionTpl> &reduced_model,
+      std::vector<GeometryModel> &list_of_reduced_geom_models) {
+    typedef ModelTpl<Scalar, Options, JointCollectionTpl> Model;
+    buildReducedModel(input_model, list_of_joints_to_lock,
+                      reference_configuration, reduced_model);
+
+    // for all GeometryModels
+    for (auto &input_geom_model : list_of_geom_models) {
```

Comment:
Don’t use auto keyword as we are still compliant with C++98

Reply:
Noted, it should be fixed in next commit.